File: skills/cumcm-code-writer/references/algorithm_library/03_预测时序/nar_mlp_forecast/matlab/main.m
% NAR 非线性自回归神经网络预测（MATLAB 最小可运行模板）
% 说明：本机无 Deep Learning Toolbox，用单隐藏层 ELM（随机隐层 + 最小二乘输出）实现同样的"滞后特征回归"。
here = fileparts(mfilename('fullpath'));
addpath(fullfile(here, '..', '..', '..', '_common', 'matlab'));

rng(42);
t = (0:119)';
series = 30 + 6*sin(2*pi*t/24) + 0.05*t + 0.4*randn(120, 1);
lags = 6; steps = 12;

X = zeros(numel(series) - lags, lags);
for k = 1:lags
    X(:, k) = series(lags - k + 1:end - k);
end
y = series(lags + 1:end);
split = round(0.8 * size(X, 1));

hidden = 16;
W = randn(lags, hidden); b = randn(1, hidden);
H = 2 ./ (1 + exp(-2 * (X * W + b))) - 1;
beta = pinv(H(1:split, :)) * y(1:split);
fitted = H * beta;
rmse = sqrt(mean((fitted(split + 1:end) - y(split + 1:end)).^2));

window = series(end - lags + 1:end);
forecast = zeros(steps, 1);
for h = 1:steps
    feat = flipud(window(end - lags + 1:end));              % lags×1
    hiddenOut = 2 ./ (1 + exp(-2 * (W' * feat + b'))) - 1;  % hidden×1
    value = beta' * hiddenOut;                              % 1×1
    forecast(h) = value;
    window = [window; value]; %#ok<AGROW>
end

mmlib_report('NAR 神经网络', sprintf('滞后阶数=%d', lags), sprintf('留出RMSE=%.4f', rmse), ...
    sprintf('预测首期=%.4f', forecast(1)), sprintf('预测末期=%.4f', forecast(end)));
T = table([y; forecast], 'VariableNames', {'数值'});
mmlib_savetable(here, 'nar_mlp_forecast', T);

% 图件由 Python 生成（技能规则：交付图必须用 Python 绘制，MATLAB 仅负责计算）

File: skills/cumcm-code-writer/references/algorithm_library/03_预测时序/nar_mlp_forecast/python/main.py
"""NAR 非线性自回归神经网络预测（Python 最小可运行模板）

思路：把时序展开成滞后特征 X = [y_{t-1}, ..., y_{t-p}]，用 MLP 拟合 y_t，再递归多步预测。
适用于非线性、非平稳但自相关明显的序列；比线性 AR 更能捕捉阈值与饱和效应。
接口：solve(series, lags, steps) -> (预测值, 拟合优度)；运行：python main.py
"""
from __future__ import annotations

import sys
from pathlib import Path

import numpy as np
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import StandardScaler

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "_common" / "python"))
from mmlib import demo, report, save_fig, save_table, set_style  # noqa: E402


def build_demo() -> dict:
    rng = np.random.default_rng(42)
    t = np.arange(120)
    series = 30 + 6 * np.sin(2 * np.pi * t / 24) + 0.05 * t + rng.normal(scale=0.4, size=120)
    return {"series": series, "lags": 6, "steps": 12}


def make_supervised(series, lags: int):
    X = np.column_stack([series[lags - k - 1: len(series) - k - 1] for k in range(lags)])
    return X, series[lags:]


def solve(series, lags: int = 6, steps: int = 12, seed: int = 42):
    series = np.asarray(series, float)
    X, y = make_supervised(series, lags)
    split = int(len(X) * 0.8)
    scaler = StandardScaler().fit(X[:split])
    model = MLPRegressor(hidden_layer_sizes=(16,), activation="tanh", solver="lbfgs",
                         max_iter=3000, random_state=seed)
    model.fit(scaler.transform(X[:split]), y[:split])
    fitted = model.predict(scaler.transform(X))
    window = list(series[-lags:])
    forecast = []
    for _ in range(steps):
        features = np.array(window[-lags:][::-1]).reshape(1, -1)
        value = float(model.predict(scaler.transform(features))[0])
        forecast.append(value)
        window.append(value)
    rmse = float(np.sqrt(np.mean((fitted[split:] - y[split:]) ** 2)))
    return np.array(forecast), fitted, y, rmse


if __name__ == "__main__":
    data = demo(__file__, build_demo)
    forecast, fitted, y, rmse = solve(data["series"], int(data["lags"]), int(data["steps"]))
    report("NAR 神经网络", 滞后阶数=int(data["lags"]), 留出RMSE=round(rmse, 4),
           预测首期=round(float(forecast[0]), 4), 预测末期=round(float(forecast[-1]), 4))
    save_table(__file__, "拟合与预测",
               {"类型": ["拟合"] * len(y) + ["预测"] * len(forecast),
                "数值": np.round(np.concatenate([y, forecast]), 4)})
    plt = set_style()
    fig, ax = plt.subplots()
    t_fit = np.arange(int(data["lags"]), int(data["lags"]) + len(y))
    t_pred = np.arange(t_fit[-1] + 1, t_fit[-1] + 1 + len(forecast))
    ax.plot(t_fit, y, color="#0072B2", label="观测值")
    ax.plot(t_fit, fitted, "--", color="#56B4E9", label="拟合值")
    ax.plot(t_pred, forecast, color="#D55E00", label="递归预测")
    ax.set_xlabel("时间")
    ax.set_ylabel("取值")
    ax.legend()
    save_fig(__file__, fig, "nar_mlp_forecast")
